function-images/tests/tuning-halving/driver/main.py

```diff
@@ -0,0 +1,262 @@
+        return key
+
+    def get(self, key):
+        msg = "Driver gets key '" + key + "' from " + self.transferType
+        log.info(msg)
+        with tracing.Span(msg):
+            response = None
+            if self.transferType == S3:
+                obj = self.s3_client.Object(bucket_name=self.benchName, key=key)
+                response = obj.get()
+            elif self.transferType == XDT:
+                log.fatal("XDT is not yet supported")
+            return pickle.loads(response['Body'].read())
+
+    def handler_broker(self, event, context):
+        dataset = generate_dataset()
+        hyperparam_config = {
+            'model': 'RandomForestRegressor',
+            'params': {
+                'n_estimators': [5, 10, 20],
+                'min_samples_split': [2, 4],
+                'random_state': [42]
+            }
+        }
+        models_config = {
+            'models': [
+                {
+                    'model': 'RandomForestRegressor',
+                    'params': hyperparam
+                } for hyperparam in generate_hyperparam_sets(hyperparam_config['params'])
+            ]
+        }
+        self.put(dataset, 'dataset_key')
+        return {
+            'dataset_key': 'dataset_key',
+            'models_config': models_config
+        }
+
+    def train(self, arg: dict) -> dict:
+        log.info("Invoke Trainer")
+        channel = grpc.insecure_channel(args.tAddr)
+        stub = tuning_pb2_grpc.TrainerStub(channel)
+
+        resp = stub.Train(tuning_pb2.TrainRequest(
+            dataset=b'',  # via S3/XDT only
+            dataset_key="dataset_key",
+            model_config=pickle.dumps(arg['model_config']),
+            count=arg['count'],
+            sample_rate=arg['sample_rate']
+        ))
+
+        return {
+            'model_key': resp.model_key,
+            'pred_key': resp.pred_key,
+            'score': resp.score,
+            'params': pickle.loads(resp.params),
+        }
+
+    # Driver code below
+    def SayHello(self, request, context):
+        log.info("Driver received a request")
+
+        event = self.handler_broker({}, {})
+        models = event['models_config']['models']
+
+        while len(models)>1:
+            sample_rate = 1/len(models)
+            log.info(f"Running {len(models)} models on the dataset with sample rate {sample_rate} ")
+            # Run different model configs on sampled dataset
+            training_responses = []
+            for count, model_config in enumerate(models):
+                training_responses.append(
+                    self.train({
+                        'dataset_key': event['dataset_key'],
+                        'model_config': model_config,
+                        'count': count,
+                        'sample_rate': sample_rate
+                    })
+                )
+
+            # Keep models with the best score
+            top_number = len(training_responses)//2
```

Comment:
is `//` a typo?

Reply:
I don't think so. I didn't type this, it's from the original code, I thought it's just integer division